ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,440 @@
 {
-	size_t carry, value;
-	unsigned char *nptr, *rptr;
+	size_t numerator_top_index = numerator_arr_size - 1;
+	size_t quot_top_index = quot_arr_size - 1;
+	for (size_t i = 0; i < quot_arr_size - 1; i++) {
+		if (numerator_vector[numerator_top_index - i] < divisor_vector) {
+			quot_vector[quot_top_index - i] = 0;
+			/* numerator_vector[numerator_top_index - i] < divisor_vector, so there will be no overflow. */
+			numerator_vector[numerator_top_index - i - 1] += numerator_vector[numerator_top_index - i] * BC_VECTOR_BOUNDARY_NUM;
+			continue;
+		}
+		quot_vector[quot_top_index - i] = numerator_vector[numerator_top_index - i] / divisor_vector;
+		numerator_vector[numerator_top_index - i] -= divisor_vector * quot_vector[quot_top_index - i];
+		numerator_vector[numerator_top_index - i - 1] += numerator_vector[numerator_top_index - i] * BC_VECTOR_BOUNDARY_NUM;
+		numerator_vector[numerator_top_index - i] = 0;
+	}
+	/* last */
+	quot_vector[0] = numerator_vector[0] / divisor_vector;
+}
 
-	if (digit == 0) {
-		memset(result, 0, size);
-	} else {
-		if (digit == 1) {
-			memcpy(result, num, size);
-		} else {
-			/* Initialize */
-			nptr = (unsigned char *) (num + size - 1);
-			rptr = (unsigned char *) (result + size - 1);
-			carry = 0;
-
-			while (size-- > 0) {
-				value = *nptr-- * digit + carry;
-				*rptr-- = value % BASE;
-				carry = value / BASE;
-			}
+/*
+ * Used when the divisor is split into 2 or more chunks.
+ */
+static inline void bc_standard_div(
+	BC_VECTOR *numerator_vector, size_t numerator_arr_size,
+	BC_VECTOR *divisor_vector, size_t divisor_arr_size, size_t divisor_len,
+	BC_VECTOR *quot_vector, size_t quot_arr_size
+) {
+	size_t numerator_top_index = numerator_arr_size - 1;
+	size_t divisor_top_index = divisor_arr_size - 1;
+	size_t quot_top_index = quot_arr_size - 1;
+
+	BC_VECTOR div_carry = 0;
+
+	/*
+	 * Errors might occur between the true quotient and the temporary quotient calculated using only the high order digits.
+	 * For example, the true quotient of 240 / 121 is 1.
+	 * However, if it is calculated using only the high-order digits (24 / 12), we would get 2.
+	 * We refer to this value of 2 as the temporary quotient.
+	 * We also define E to be error between the true quotient and the temporary quotient,
+	 * which in this case, is 1.
+	 * Also note that for example 2400000 / 120, there will be 5 divisions.
+	 *
+	 * Another example: Calculating 99900000000 / 10009999 with base 10000.
+	 * The actual quotient is 9980, but if calculate it using only the high-order digits:
+	 * 999 / 1000 = 0, Numbers that are not divisible are carried forward to the next division.
+	 * 9990000 / 1000 = 9990 (Here, the temporary quotient is 9990 and the error E is 10.)
+	 *
+	 * If calculate the temporary quotient using only one array of numerator and divisor, the error E can be larger than 1.
+	 * In other words, in the restoring division, the count of additions for restore increases significantly.
```

Comment:
I am not exactly sure what you mean by
> using only one array for numerator and divisor

Reply:
The other thing, which I would like to know, is _how_ much of an impact in performance do the repeated restores actual cause.

Because if it is negligible, I would just not bother, it would also make the initial implementation easier to review.

Reply:
> I am not exactly sure what you mean by

This means that an error will occur if calculations are made using only the high parts.

> The other thing, which I would like to know, is how much of an impact in performance do the repeated restores actual cause.
> 
> Because if it is negligible, I would just not bother, it would also make the initial implementation easier to review.

I'll try measuring it.

Reply:
```suggestion
	 * Because we use the restoring division we need to perform E restorations, which can be significant if E is large.
```

Feels like it conveys the same meaning, just way more succinctly

Reply:
@Girgias 
Measurements were made under conditions that would most likely make a difference.

code:
```
$num1 = '90000000000000000000000000000000000000000000000000000000000000';
$num2 = '109999999999999999999999999999999999999999';

$start = microtime(true);
for ($i = 0; $i < 500; $i++) {
    bcdiv($num1, $num2, 0);
}
echo microtime(true) - $start . PHP_EOL;
```

Multiple restores: 13.932011127472s
This PR: 0.00018405914306641s

Reply:
Fixed the comment
https://github.com/php/php-src/pull/14660/commits/72f7f3a1c7f810fa6495ecf09f6e6f6b3d94d6c9

Reply:
> @Girgias Measurements were made under conditions that would most likely make a difference.
> 
> code:
> 
> ```
> $num1 = '90000000000000000000000000000000000000000000000000000000000000';
> $num2 = '109999999999999999999999999999999999999999';
> 
> $start = microtime(true);
> for ($i = 0; $i < 500; $i++) {
>     bcdiv($num1, $num2, 0);
> }
> echo microtime(true) - $start . PHP_EOL;
> ```
> 
> Multiple restores: 13.932011127472s This PR: 0.00018405914306641s

Do you have the implementation without a restore in a branch somewhere just so that I can have a look?

Reply:
@Girgias 
Made a PR in my repository :)
https://github.com/SakiTakamachi/php-src/pull/10